plugins/scaffolder-backend/src/ScaffolderPlugin.ts

```diff
@@ -23,120 +23,104 @@ import { ScmIntegrations } from '@backstage/integration';
+  TaskBroker,
   TemplateAction,
-} from '@backstage/plugin-scaffolder-node';
-import {
   TemplateFilter,
   TemplateGlobal,
-  TaskBroker,
-} from '@backstage/plugin-scaffolder-backend';
+} from '@backstage/plugin-scaffolder-node';
 import { createBuiltinActions } from './scaffolder';
 import { createRouter } from './service/router';
 
 /**
- * Catalog plugin options
+ * Scaffolder plugin
  *
  * @alpha
  */
-export type ScaffolderPluginOptions = {
-  actions?: TemplateAction<any, any>[];
-  taskWorkers?: number;
-  taskBroker?: TaskBroker;
-  additionalTemplateFilters?: Record<string, TemplateFilter>;
-  additionalTemplateGlobals?: Record<string, TemplateGlobal>;
-};
+export const scaffolderPlugin = createBackendPlugin({
+  pluginId: 'scaffolder',
+  register(env) {
+    const addedActions = new Array<TemplateAction<any, any>>();
+    env.registerExtensionPoint(scaffolderActionsExtensionPoint, {
+      addActions(...newActions: TemplateAction<any>[]) {
+        addedActions.push(...newActions);
+      },
+    });
 
-class ScaffolderActionsExtensionPointImpl
-  implements ScaffolderActionsExtensionPoint
-{
-  #actions = new Array<TemplateAction<any, any>>();
+    let taskBroker: TaskBroker | undefined;
+    env.registerExtensionPoint(scaffolderTaskBrokerExtensionPoint, {
+      setTaskBroker(newTaskBroker) {
+        if (taskBroker) {
+          throw new Error('Task broker may only be set once');
+        }
+        taskBroker = newTaskBroker;
+      },
+    });
 
-  addActions(...actions: TemplateAction<any>[]): void {
-    this.#actions.push(...actions);
-  }
+    const additionalTemplateFilters: Record<string, TemplateFilter> = {};
+    const additionalTemplateGlobals: Record<string, TemplateGlobal> = {};
+    env.registerExtensionPoint(scaffolderTemplatingExtensionPoint, {
+      addTemplateFilters(newFilters) {
+        Object.assign(additionalTemplateFilters, newFilters);
+      },
+      addTemplateGlobals(newGlobals) {
+        Object.assign(additionalTemplateGlobals, newGlobals);
+      },
+    });
 
-  get actions() {
-    return this.#actions;
-  }
-}
+    env.registerInit({
+      deps: {
+        logger: coreServices.logger,
+        config: coreServices.rootConfig,
+        reader: coreServices.urlReader,
+        permissions: coreServices.permissions,
+        database: coreServices.database,
+        httpRouter: coreServices.httpRouter,
+        catalogClient: catalogServiceRef,
+      },
+      async init({
+        logger,
+        config,
+        reader,
+        database,
+        httpRouter,
+        catalogClient,
+        permissions,
+      }) {
+        const log = loggerToWinstonLogger(logger);
 
-/**
- * Catalog plugin
- *
- * @alpha
- */
-export const scaffolderPlugin = createBackendPlugin(
-  (options?: ScaffolderPluginOptions) => ({
-    pluginId: 'scaffolder',
-    register(env) {
-      const actionsExtensions = new ScaffolderActionsExtensionPointImpl();
+        const actions = [
+          ...addedActions,
```

Comment:
no overriding here? as in, should these be spread in last or first, or doesn't it matter

Reply:
Kept it as it was, but was also a bit curious about what makes sense here in practice. Thinking we could tweak it in the future if needed. The main thing I was worried about was not being able to disable the built-in actions, it does feel a bit awkward to add that as an extension point though. Maybe a `scaffolder.disableBuildingActions` config is the way to go?

Reply:
As with any "collection like" thing I think it's the most robust if we always go for complete replacement but with facilities for easily reaching the builtins. Same as with table columns etc, lots of prior art there